rust/src/router.rs

```diff
@@ -1,38 +1,90 @@
+use crate::tree::RadixTree;
 use actix_web::http::header::{HeaderValue, CONTENT_TYPE};
 use actix_web::{HttpRequest, HttpResponse};
 use bytes::Bytes;
 use futures_util::TryStreamExt;
+use std::collections::HashMap;
 use std::fmt::Debug;
+use std::sync::atomic::AtomicUsize;
+use std::sync::{Arc, Mutex};
+use tokenizers::tokenizer::Tokenizer;
 
 #[derive(Debug)]
 pub enum Router {
     RoundRobin {
         worker_urls: Vec<String>,
-        current_index: std::sync::atomic::AtomicUsize,
+        current_index: AtomicUsize,
     },
     Random {
         worker_urls: Vec<String>,
     },
+    ApproxTree {
+        worker_urls: Vec<String>,
```

Comment:
If `worker_urls` is **mostly for read only**, instead of owned string, can use `&str` or `Arc<str>` for thread-safe. Consider `Vec<Arc<str>>` in your case

Reply:
For now Vec<String> is read-only, but once we add dynamic scaling, it will be read+write and mostly read. How would you suggest `Vec<Arc<String>` v.s.  `Vec<Arc<str>>` (why not `Vec<Arc<&str>> btw?)